detectable/src/main/java/com/synopsys/integration/detectable/detectables/gradle/inspection/parse/GradleReportTransformer.java

```diff
@@ -51,35 +57,57 @@ public CodeLocation transform(GradleReport gradleReport) {
 
-        for (final GradleTreeNode currentNode : configuration.getChildren()) {
-
-            if (skipUntil.isPresent() && currentNode.getLevel() <= skipUntil.get()) {
-                skipUntil = Optional.empty();
-            } else if (skipUntil.isPresent()) {
+        TreeNodeSkipper treeNodeSkipper = null;
+        for (GradleTreeNode currentNode : configuration.getChildren()) {
+            if (treeNodeSkipper != null && treeNodeSkipper.shouldSkip(currentNode)) {
                 continue;
+            } else if (treeNodeSkipper != null) {
+                // TreeNodeSkipper has stopped skipping lines.
+                treeNodeSkipper = null;
             }
 
-            history.clearDependenciesDeeperThan(currentNode.getLevel());
-            if (currentNode.getNodeType() != GradleTreeNode.NodeType.GAV) {
-                skipUntil = Optional.of(currentNode.getLevel());
-                continue;
+            if (currentNode.getNodeType() == GradleTreeNode.NodeType.GAV) {
+                history.clearDependenciesDeeperThan(currentNode.getLevel());
+                Optional<GradleGav> currentNodeGav = currentNode.getGav();
+                if (currentNodeGav.isPresent()) {
+                    addGavToGraph(currentNodeGav.get(), history, graph);
+                } else {
+                    // We know this is a GradleTreeNode.NodeType.GAV
+                    // So if its missing data, something is probably wrong.
+                    logger.debug("Missing expected GAV from known NodeType. {}", currentNode);
+                }
+            } else {
+                treeNodeSkipper = new TreeNodeSkipper(currentNode);
             }
+        }
+    }
 
-            final GradleGav gav = currentNode.getGav().get(); // TODO: Why are we not doing an isPresent() check here?
-            final ExternalId externalId = externalIdFactory.createMavenExternalId(gav.getName(), gav.getGroup(), gav.getVersion());
-            final Dependency currentDependency = new Dependency(gav.getGroup(), gav.getVersion(), externalId);
+    private void addGavToGraph(GradleGav gav, DependencyHistory history, MutableDependencyGraph graph) {
+        ExternalId externalId = externalIdFactory.createMavenExternalId(gav.getName(), gav.getGroup(), gav.getVersion());
+        Dependency currentDependency = new Dependency(gav.getGroup(), gav.getVersion(), externalId);
 
-            if (history.isEmpty()) {
-                graph.addChildToRoot(currentDependency);
-            } else {
-                graph.addChildWithParents(currentDependency, history.getLastDependency());
-            }
-            history.add(currentDependency);
+        if (history.isEmpty()) {
+            graph.addChildToRoot(currentDependency);
+        } else {
+            graph.addChildWithParents(currentDependency, history.getLastDependency());
+        }
+        history.add(currentDependency);
+    }
+
+    private static class TreeNodeSkipper {
```

Comment:
The way I see it you have two options: 

The first option: The state lives in the transformer, and in that case I prefer the old code (or at least the old code with improvements). I don't see a benefit to abstracting the line level check. There is only one implementation. It adds these null checks. The code is further away and as far as I can tell I would need to read both to understand the block anyway. And it doesn't really simplify the transformers code. 

The second option: You move the state to the TreeSkipper to simplify the transformer. I think the argument is there is too much happening in the transformer and you want to simplify. To do that I think it makes sense to move some of the state to the TreeSkipper.

This simplifies the transformer a bit: 
```
        TreeNodeSkipper treeNodeSkipper = new TreeNodeSkipper();
        for (GradleTreeNode currentNode : configuration.getChildren()) {
            if (treeNodeSkipper.shouldSkip(currentNode)) {
                continue;
            }

            if (currentNode.getNodeType() == GradleTreeNode.NodeType.GAV) {
                ...
            } else {
                treeNodeSkipper.skipUntilLineLevel(currentNode.getLineLevel());
            }
        }
```

But at the cost of adding a bit of state and complexity to the TreeSkipper.

```
private static class TreeNodeSkipper {
            private Optional<Integer> skipUntilLineLevel = Optional.empty();

            public boolean shouldSkip(GradleTreeNode nodeInQuestion) {
                if (skipUntilLineLevel.isPresent()) {
                    if (nodeInQuestion.getLevel() > skipUntilLineLevel) {
                        return true;
                    } else {
                        skipUntilLineLevel = Optional.empty();
                        return false;
                    }
                } else {
                    return false;
                }
            }

            public void skipUntilLineLevel(int lineLevel) {
                skipUntilLineLevel = Optional.of(lineLevel);
            }
    }
``` 

I think in this case you still have that distance but hopefully the transformer is now easier to follow. Whether that's worth adding distance is I suppose in the eye of the beholder. I bet there are some improvements that could be made to the original code to make it clearer without adding all this distance but I don't mind it if we want to go a TreeSkip route.

That's my 2 cents.